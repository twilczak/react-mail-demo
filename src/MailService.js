import axios from 'axios';

export class MailService {

    static hostUrl = 'http://localhost:3000';

    static getMessages(mailbox) {
        const url = `${this.hostUrl}/` + mailbox.split('/').join('');
        return axios
            .get(url)
            .catch(this.handleError);
    }

    static getMessage(mailbox, id) {
        const url = `${this.hostUrl}/${mailbox}`;
        return axios
            .get(url)
            .then(response => {
                return response.data.find(message => message.id === id);
            })
            .catch(this.handleError);
    }

    static deleteMessage(mailbox, id) {
        const url = `${this.hostUrl}/${mailbox}/${id}`;
        return axios
            .delete(url)
            .catch(this.handleError);
    }

    static sendMessage(message) {
        const url = `${this.hostUrl}/outbox`;
        message.dateSent = this.getDateSent();

        return axios
            .post(url, message)
            .then(response => {
                console.log(response);
                return response.data;
            })
            .catch(this.handleError);
    }

    static zeroPad(value) {
        return value > 9 ?  '' + value : '0' + value;
    }

    static getDateSent(date = new Date()) {
        const day = this.zeroPad(date.getDate());
        const month = this.zeroPad(date.getMonth() + 1);
        const year = date.getFullYear();

        return `${year}.${month}.${day}`;
    }

    static handleError(error) {
        console.error(error);
    }
}