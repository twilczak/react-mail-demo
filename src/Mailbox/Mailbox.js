import React, { Component } from 'react';

import './Mailbox.css';

import { MailService } from "../MailService";

import { MailboxControls } from "./MailboxControls/MailboxControls";
import { MailboxList } from "./MailboxList/MailboxList";

export class Mailbox extends Component {

    constructor(props) {
        super(props);

        this.state = {
            messages: []
        };
    }

    componentDidMount() {
        this.loadMessages(this.props.match.path);
    }

    componentWillReceiveProps({match}) {
        this.loadMessages(match.path);
    }

    loadMessages(path) {
        this.setState({messages: []});
        MailService
            .getMessages(path)
            .then((response) => this.setState({messages: response.data}));
    }

    render() {
        return (
            <section className="mailbox">

                <MailboxControls match={this.props.match}/>
                <MailboxList match={this.props.match} messages={this.state.messages}/>

                <hr className="mailbox-border"/>

            </section>
        );
    }
}