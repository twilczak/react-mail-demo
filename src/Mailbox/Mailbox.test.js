import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { MemoryRouter, Route } from 'react-router-dom';
import { Mailbox } from "./Mailbox";

import { MailService } from "../MailService";

jest.mock('../MailService');

describe(Mailbox, () => {
    const mockMessages = [{id: 'test'}];

    MailService.getMessages.mockReturnValue(
        { then: (callback) => {callback({data: mockMessages});} }
    );

    const component = mount(
        <MemoryRouter initialEntries={['/inbox']}>
            <Route path="/inbox" component={Mailbox}/>
        </MemoryRouter>
    );

    it('renders and matches the snapshot', () => {
        const component = renderer.create(
            <MemoryRouter initialEntries={['/inbox']}>
                <Route path="/inbox" component={Mailbox}/>
            </MemoryRouter>
        );
        expect(component).toMatchSnapshot();
    });

    it('renders the mailbox markup', () => {
        expect(component.find('.mailbox')).toBeTruthy();
        expect(component.find('MailboxControls')).toBeTruthy();
        expect(component.find('MailboxList')).toBeTruthy();
        expect(component.find('.mailbox-border')).toBeTruthy();
    });

    it('calls MailService.getMessages()', () => {
        expect(MailService.getMessages).toBeCalledWith('/inbox');
    });
});
