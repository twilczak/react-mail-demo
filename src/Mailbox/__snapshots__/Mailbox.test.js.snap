// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Mailbox extends _react.Component {

    constructor(props) {
        super(props);

        this.state = {
            messages: [] };

    }

    componentDidMount() {
        this.loadMessages(this.props.match.path);
    }

    componentWillReceiveProps({ match }) {
        this.loadMessages(match.path);
    }

    loadMessages(path) {
        this.setState({ messages: [] });
        _MailService.MailService.
        getMessages(path).
        then(({ data }) => this.setState({ messages: data }));
    }

    render() {
        return (
            _react2.default.createElement('section', { className: 'mailbox', __source: { fileName: _jsxFileName, lineNumber: 40 }, __self: this },

                _react2.default.createElement(_MailboxControls.MailboxControls, { match: this.props.match, __source: { fileName: _jsxFileName, lineNumber: 42 }, __self: this }),
                _react2.default.createElement(_MailboxList.MailboxList, { match: this.props.match, location: this.props.location, messages: this.state.messages, __source: { fileName: _jsxFileName, lineNumber: 43 }, __self: this }),

                _react2.default.createElement('hr', { className: 'mailbox-border', __source: { fileName: _jsxFileName, lineNumber: 45 }, __self: this }),

                _react2.default.createElement(_reactRouterDom.Switch, { __source: { fileName: _jsxFileName, lineNumber: 47 }, __self: this },
                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/inbox/view/:messageId', component: _MessageReader.MessageReader, __source: { fileName: _jsxFileName, lineNumber: 48 }, __self: this }),
                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/outbox/view/:messageId', component: _MessageReader.MessageReader, __source: { fileName: _jsxFileName, lineNumber: 49 }, __self: this }),

                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/inbox/compose', component: _MessageComposer.MessageComposer, __source: { fileName: _jsxFileName, lineNumber: 51 }, __self: this }),
                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/outbox/compose', component: _MessageComposer.MessageComposer, __source: { fileName: _jsxFileName, lineNumber: 52 }, __self: this }))));




    }} renders and matches the snapshot 1`] = `
<section
  className="mailbox"
>
  <div>
    <a
      aria-current="true"
      className="mailbox-button is-active"
      href="/inbox"
      onClick={[Function]}
      style={Object {}}
    >
      Inbox
    </a>
    <a
      aria-current={false}
      className="mailbox-button"
      href="/outbox"
      onClick={[Function]}
      style={undefined}
    >
      Outbox
    </a>
    <a
      className="mailbox-button"
      href="/inbox/compose"
      onClick={[Function]}
    >
      Compose
    </a>
  </div>
  <ol
    className="mailbox-list"
  >
    <li
      className="mailbox-list-header"
    >
      <span
        className="sender"
      >
        Sender
      </span>
      <span
        className="subject"
      >
        Subject
      </span>
      <span
        className="date-sent"
      >
        Date
      </span>
    </li>
    <li
      className="mailbox-list-item"
      id="test"
    >
      <a
        aria-current={false}
        className="mailbox-item-link"
        href="/inbox/view/test"
        onClick={[Function]}
        style={undefined}
      >
        <span
          className="sender"
        />
        <span
          className="subject"
        />
        <span
          className="date-sent"
        />
      </a>
    </li>
  </ol>
  <hr
    className="mailbox-border"
  />
</section>
`;
